lms/djangoapps/discussion/rest_api/api.py

```diff
@@ -1250,3 +1255,84 @@ def delete_comment(request, comment_id):
         comment_deleted.send(sender=None, user=request.user, post=cc_comment)
     else:
         raise PermissionDenied
+
+
+@function_trace("_get_course_discussion_user_stats")
+def _get_course_discussion_user_stats(
+    course_key: CourseKey,
+    order_by: UserOrdering = UserOrdering.BY_FLAGS,
+) -> List[Dict[str, Union[str, int]]]:
+    """
+    Get course user discussion stats.
+
+    Args:
+        course_key (CourseKey): course key for course to fetch discussion stats for.
+        order_by (UserOrdering): the ordering to use for sorting data
+
+    Returns:
+        Sorted user stat data for the specified course
+    """
+    course_stats = get_course_user_stats(course_key)
+
+    def key_func(user):
+        key = (user["threads"], user["responses"], user["replies"])
+        if order_by == UserOrdering.BY_FLAGS:
+            key = (user["active_flags"], user["inactive_flags"]) + key
+        return key
+
+    return sorted(course_stats.values(), key=key_func, reverse=True)
+
+
+@function_trace("get_course_discussion_user_stats")
+def get_course_discussion_user_stats(
+    request,
+    course_key_str: str,
+    page: int,
+    page_size: int,
+    order_by: UserOrdering = None,
+) -> Dict:
+    """
+    Get paginated course discussion stats for users in the course.
+
+    Args:
+        request (Request): DRF request
+        course_key_str (str): course key string
+        page (int): Page number to fetch
+        page_size (int): Number of items in each page
+        order_by (UserOrdering): The ordering to use for the user stats
+
+    Returns:
+        Paginated data of a user's discussion stats sorted based on the specified ordering.
+
+    """
+    course_key = CourseKey.from_string(course_key_str)
+    if has_discussion_privileges(user=request.user, course_id=course_key):
+        order_by = order_by or UserOrdering.BY_FLAGS
+    else:
+        order_by = order_by or UserOrdering.BY_ACTIVITY
+        if order_by != UserOrdering.BY_ACTIVITY:
+            raise ValidationError({"order_by": ["Invalid value"]})
+    data = _get_course_discussion_user_stats(course_key, order_by)
+    num_items = len(data)
+    num_pages = ceil(num_items / page_size)
+    start_idx = (page - 1) * page_size
+    data = data[start_idx:start_idx + page_size]
+    user_ids = [user["author_id"] for user in data]
+    if not data and page != 1:
+        raise PageNotFoundError("Page not found (No results on this page).")
+    serializer = UserStatsSerializer(
+        data,
+        many=True,
+        context={
+            "users": dict(User.objects.filter(id__in=user_ids).values_list("id", "username"))
+        }
```

Comment:
Why this context is needed in `UserStatsSerializer`?

Reply:
There can be other ways to simplify that but another way could be (coming into my mind right now) - this can remove the deps of user context for `UserStatsSerializer`. 
```
data = ....
users_data = User.objects.filter(id__in=user_ids).values_list("id", "username")
data_with_usernames = [
  { **stats, "username": users_data[stats['author_id']] } 
  for stats in data
]
```

Reply:
I'm not sure of the advantage of doing this though? Why is passing this as context a bad idea?

Reply:
It just simplifies the code and removes the deps for providing user context to `UserStatsSerializer`. Given that it'll be easier to test as well. 
I am not proposing the above code to follow but an option -  I wanted to add that it can be achieved without passing the data as context (shared some adv above). 

I am ok with whatever you decide is best. 👍

Reply:
I've added the change you suggested, but I think overall it doesn't really simplify the code. Instead of iterating over the data twice (once to get user ids and once for the serialiser) we now do it three times (get user ids, add usernames to data, and then serialise). I've put it in a separate commit so if you can see the change and it can be easily reverted if needed.

Reply:
thanks I'll take a look.